src/js/formers/SaleFormer.js

```diff
@@ -0,0 +1,199 @@
+    this.attrs.isWhitelisted = source.isWhitelisted
+    this.attrs.saleLogo = source.logo
+    this.attrs.saleLogo.key = source.logoKey
+    this.attrs.shortDescription = source.shortDescription
+    this.attrs.youtubeVideo = source.youtubeVideoId || ''
+    this.attrs.fullDescription = source.description || ''
+  }
+
+  async buildOps () {
+    await Document.uploadDocuments([this.attrs.saleLogo])
+    return [await this._buildOp()]
+  }
+
+  async _buildOp () {
+    this.attrs.saleDescriptionBlobId = await createSaleDescriptionBlob(
+      this.attrs.fullDescription,
+      this.attrs.creatorAccountId
+    )
+
+    await createBalancesIfNotExist({
+      balanceAssets: this.attrs.assets.map(asset => asset.code),
+      quoteAssets: this.attrs.quoteAssets,
+      accountId: this.attrs.creatorAccountId,
+    })
+    this.attrs.assetPairs = await this._loadAssetsPairsByQuote(
+      this.attrs.capAssetCode
+    )
+
+    const opts = this._createSaleRequestOpts()
+    return base.SaleRequestBuilder.createSaleCreationRequest(opts)
+  }
+
+  _createSaleRequestOpts () {
+    const saleLogo = this.attrs.saleLogo
+    const defaultSaleType = '0'
+
+    return {
+      requestID: this.attrs.requestId || '0',
+      saleEnumType: this.attrs.saleType,
+      saleType: defaultSaleType,
+      startTime: DateUtil.toTimestamp(this.attrs.startTime),
+      endTime: DateUtil.toTimestamp(this.attrs.endTime),
+      baseAsset: this.attrs.baseAssetCode,
+      defaultQuoteAsset: this.attrs.capAssetCode,
+      softCap: this.attrs.softCap,
+      hardCap: this.attrs.hardCap,
+      requiredBaseAssetForHardCap: this.attrs.assetsToSell,
+      quoteAssets: this._getQuoteAssets(),
+      creatorDetails: {
+        name: this.attrs.saleName,
+        short_description: this.attrs.shortDescription,
+        description: this.attrs.saleDescriptionBlobId,
+        logo: saleLogo || new Document(),
+        youtube_video_id: this.attrs.youtubeVideo,
+      },
+      saleRules: [{
+        forbids: this.attrs.isWhitelisted,
+      }],
+    }
+  }
+
+  _getQuoteAssets () {
+    const basePrise = MathUtil.divide(
+      this.attrs.hardCap,
+      this.attrs.assetsToSell
+    )
+
+    let a = this.attrs.quoteAssets.map((item) => ({
+      asset: item,
+      price: this._getPrice(item, basePrise),
+    }))
+    return a
+  }
+
+  _getPrice (assetCode, basePrise) {
+    let result
+
+    const capAsset = this.attrs.capAssetCode
+    if (capAsset !== assetCode) {
+      if (this.attrs.saleType === SALE_TYPES.immediate) {
+        let assetPair = this.attrs.assetPairs.filter(item =>
+          item.baseAndQuote === `${assetCode}/${capAsset}`
+        )
+        result = MathUtil.divide(basePrise, assetPair[0].price)
+      } else {
+        const defaultQuoteAssetPrice = '1'
+        result = defaultQuoteAssetPrice
+      }
+    } else {
+      result = basePrise
+    }
+
+    return result
+  }
+
+  async _loadAssetsPairsByQuote (quoteAssetCode) {
```

Comment:
Formers is our implementation of the "Constructor" pattern. It should collect input and simply give you the output, the only reason for it to exist: allow the programmer to collect the input step-by-step. In the case when it tries to load something by itself, you should handle the errors when on `buildOps` call, that's definitely not a good thing.

I would recommend adding a "useAssetPairs: method instead that accepts the asset pairs and saves their anywhere but in `attrs` (cuz in this case asset pairs is support data, `attrs` is used for the collected input), thus you can call the new method directly before the `buildOps`